cli/src/state/hooks/useApprovalHandler.ts

```diff
@@ -284,13 +292,108 @@ export function useApprovalHandler(): UseApprovalHandlerReturn {
 		[store, sendAskResponse, approvalTelemetry],
 	)
 
+	const sendTerminalOperation = useCallback(
+		async (operation: "continue" | "abort") => {
+			// Read the current state directly from the store at call time
+			const currentPendingApproval = store.get(pendingApprovalAtom)
+			const processingState = store.get(approvalProcessingAtom)
+
+			if (!currentPendingApproval) {
+				logs.warn("No pending approval for terminal operation", "useApprovalHandler", {
+					operation,
+				})
+				return
+			}
+
+			// Verify this is a command_output ask
+			if (currentPendingApproval.ask !== "command_output") {
+				logs.warn("Terminal operation called for non-command_output ask", "useApprovalHandler", {
+					ask: currentPendingApproval.ask,
+					operation,
+				})
+				return
+			}
+
+			// Check if already processing
+			if (processingState.isProcessing) {
+				logs.warn("Approval already being processed, skipping duplicate", "useApprovalHandler", {
+					processingTs: processingState.processingTs,
+					currentTs: currentPendingApproval.ts,
+				})
+				return
+			}
+
+			// Start processing atomically
+			const started = store.set(startApprovalProcessingAtom, operation === "continue" ? "approve" : "reject")
+			if (!started) {
+				logs.warn("Failed to start terminal operation processing", "useApprovalHandler")
+				return
+			}
+
+			try {
+				logs.debug(`Sending terminal operation: ${operation}`, "useApprovalHandler", {
+					ts: currentPendingApproval.ts,
+				})
+
+				// Mark message as answered locally BEFORE sending operation
+				const answeredMessage: ExtensionChatMessage = {
+					...currentPendingApproval,
+					isAnswered: true,
+				}
+				store.set(updateChatMessageByTsAtom, answeredMessage)
+
+				// For terminal operations, we only need to send the terminal operation message
+				// DO NOT send askResponse - that would resolve the wrong ask (completion_result instead of command_output)
+				// The backend's onLine callback handles the command_output ask internally
+				await sendMessage({
+					type: "terminalOperation",
+					terminalOperation: operation,
+				})
+
+				logs.debug("Terminal operation sent successfully", "useApprovalHandler", {
+					operation,
+					ts: currentPendingApproval.ts,
+				})
+
+				// Track the operation as manual approval/rejection
+				if (operation === "continue") {
+					approvalTelemetry.trackManualApproval(currentPendingApproval)
+				} else {
+					approvalTelemetry.trackManualRejection(currentPendingApproval)
+				}
+
+				// Complete processing atomically
+				store.set(completeApprovalProcessingAtom)
+			} catch (error) {
```

Comment:
Ok apologies in advanced - potentially dumb question, but sendTerminalOperation sets isAnswered=true and updates the chat message locally, and then calls sendMessage. If sendMessage throws, the catch only clears the processing atom, I think?

Does the chat entry stay isAnswered: true? If so, is that an issue?

Reply:
I should be test this but, in theory the extension continue pushing the state as `isAnswered: false` and clearing the `completeApprovalProcessingAtom` will be force to reupdate the ui showing the approval menu again. 
There are multiples race conditions that this code handle on the CLI to deal with the extension message system.